geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/filters/CqlTransformFilter.scala

```diff
@@ -0,0 +1,346 @@
+        val filter = deserialize(key.bytes)
+        logger.trace(s"Deserialized $filter")
+        filter
+      }
+    }
+  )
+
+  /**
+    * Override of static method from org.apache.hadoop.hbase.filter.Filter
+    *
+    * @param pbBytes serialized bytes
+    * @throws org.apache.hadoop.hbase.exceptions.DeserializationException serialization exception
+    * @return
+    */
+  @throws(classOf[DeserializationException])
+  def parseFrom(pbBytes: Array[Byte]): org.apache.hadoop.hbase.filter.Filter =
+    // note: we copy KryoBufferSimpleFeatures as they seem to cause problems if re-used, even if thread-local
+    new CqlTransformFilter(cache.get(new ByteArrayCacheKey(pbBytes)).copy(), pbBytes)
+
+  /**
+    * Create a new filter. Typically, filters created by this method will just be serialized to bytes and sent
+    * as part of an HBase scan
+    *
+    * Note: calling this without at least one of the cql filter or transform would introduce extra processing
+    * but not actually do anything, and thus is not allowed
+    *
+    * @param sft simple feature type
+    * @param filter cql filter, if any
+    * @param transform transform, if any
+    * @return
+    */
+  def apply(sft: SimpleFeatureType,
+            filter: Option[Filter],
+            transform: Option[(String, SimpleFeatureType)]): CqlTransformFilter = {
+    if (filter.isEmpty && transform.isEmpty) {
+      throw new IllegalArgumentException("The filter must have a predicate and/or transform")
+    }
+
+    val feature = KryoFeatureSerializer(sft, SerializationOptions.withoutId).getReusableFeature
+    transform.foreach { case (tdefs, tsft) => feature.setTransforms(tdefs, tsft) }
+
+    val delegate = filter match {
+      case None => new TransformDelegate(sft, feature)
+      case Some(f) if transform.isEmpty => new FilterDelegate(sft, feature, f)
+      case Some(f) => new FilterTransformDelegate(sft, feature, f)
+    }
+
+    new CqlTransformFilter(delegate, serialize(delegate))
+  }
+
+  /**
+    * Full serialization of the delegate filter
+    *
+    * @param delegate delegate filter
+    * @return
+    */
+  private def serialize(delegate: DelegateFilter): Array[Byte] = {
+    val sftBytes = SimpleFeatureTypes.encodeType(delegate.sft, includeUserData = true).getBytes(StandardCharsets.UTF_8)
+    val cqlBytes = delegate.filter.map(ECQL.toCQL(_).getBytes(StandardCharsets.UTF_8)).getOrElse(Array.empty)
+    delegate.transform match {
+      case None =>
+        val array = Array.ofDim[Byte](sftBytes.length + cqlBytes.length + 12)
+
+        var offset = 0
+        ByteArrays.writeInt(sftBytes.length, array, offset)
+        offset += 4
+        System.arraycopy(sftBytes, 0, array, offset, sftBytes.length)
+        offset += sftBytes.length
+        ByteArrays.writeInt(cqlBytes.length, array, offset)
+        offset += 4
+        System.arraycopy(cqlBytes, 0, array, offset, cqlBytes.length)
+        offset += cqlBytes.length
+        ByteArrays.writeInt(0, array, offset)
```

Comment:
Is this to indicate that the length of the transforms is 0?

Reply:
right, I'll add a comment